### PR TITLE
Remove redundant serialization and deserialization in printer

### PR DESCRIPTION
**Pull request by icsaszar**
Affects #128.

- [ ] Needs tests.

### CODE REVIEW

**Comment by ebkalderon** [comment]
Changes look good overall! Continuing the discussion from #108, I wonder what we should do about the out-of-order serialization of fields. It's not a problem to simply update the failing tests, in theory, but now I realize that the serialization order of request fields is now inconsistent between client-to-server messages and server-to-client messages, since they aren't relying on the same `jsonrpc_core::types::params::Params` implementation of `Serialize` anymore.

File: src/delegate/printer.rs [hunk 1]
```diff
@@ -162,12 +162,17 @@ where
...
+        params: T,
+        id: Id,
+    }
     // Since these types come from the `lsp-types` crate and validity is enforced via the
     // `Request` trait, the `unwrap()` calls below should never fail.
-    let output = serde_json::to_string(&params).unwrap();
-    let params = serde_json::from_str(&output).unwrap();
-    serde_json::to_string(&request::MethodCall {
-        jsonrpc: Some(Version::V2),
+    serde_json::to_string(&RawRequest::<N::Params> {
```
Comment (ebkalderon) [inline src/delegate/printer.rs]: Is this turbofish necessary when the `make_request()` signature already contains `params: N::Params`?

**Comment by ebkalderon** [comment]
An alternative approach I just thought of which still has a two-step serialization process, but reuses the upstream `jsonrpc` type for consistent field ordering, is much lighter weight than the status quo, and avoids the double string allocation is to use [`serde_json::to_value()`](https://docs.serde.rs/serde_json/fn.to_value.html) and [`serde_json::from_value()`](https://docs.serde.rs/serde_json/fn.from_value.html) to convert from an `N::Params` to a `jsonrpc_core::types::params::Params`.

**Comment by ebkalderon** [comment]
Okay, I've opened #131 which would unblock us from working on refactoring `Printer` and adding support for proper client requests. Let's review this PR first and then move on to that one, since you opened this first.

BTW, I also merged a tiny logging PR in the meantime, so I'm sorry for rendering your PR out of date. Would you mind kindly rebasing your PR against `master`? Feel free to ping me if you need anything.

**Comment by icsaszar** [comment]
@ebkalderon What do you think of testing the current (custom struct) implementation with deserialinzing the strings into maps and comparing them? I've pushed a commit that changes the tests to compare the parsed `jsonrpc_core::Value`.

But if you would prefer to use `to_value` and `from_value` instead of the current solution I'll make the change to that.

File: src/delegate/printer.rs [hunk 2]
```diff
@@ -181,12 +186,16 @@ where
...
+    }
     // Since these types come from the `lsp-types` crate and validity is enforced via the
     // `Notification` trait, the `unwrap()` calls below should never fail.
-    let output = serde_json::to_string(&params).unwrap();
-    let params = serde_json::from_str(&output).unwrap();
-    serde_json::to_string(&request::Notification {
-        jsonrpc: Some(Version::V2),
+    serde_json::to_string(&RawNotification {
+        jsonrpc: Version::V2,
         method: N::METHOD.to_owned(),
```
Comment (ebkalderon) [inline src/delegate/printer.rs]: By changing the `method` field to `&'static str`, we can eliminate this `to_owned()` call.

File: src/delegate/printer.rs [hunk 3]
```diff
@@ -162,12 +162,17 @@ where
...
     // Since these types come from the `lsp-types` crate and validity is enforced via the
     // `Request` trait, the `unwrap()` calls below should never fail.
-    let output = serde_json::to_string(&params).unwrap();
-    let params = serde_json::from_str(&output).unwrap();
-    serde_json::to_string(&request::MethodCall {
-        jsonrpc: Some(Version::V2),
+    serde_json::to_string(&RawRequest {
+        jsonrpc: Version::V2,
         id: Id::Num(id),
         method: N::METHOD.to_owned(),
```
Comment (ebkalderon) [inline src/delegate/printer.rs]: By changing the `method` field to `&'static str`, we can eliminate this `to_owned()` call.

**Comment by ebkalderon** [comment]
Also, since #131 has been merged, please consider rebasing your PR against the latest `master`! The changes mostly affected `src/delegate.rs` and not `src/delegate/printer.rs`, so this should hopefully have no conflicts. As always, feel free to ping me if you run into issues. ☺️ 

After this PR gets merged, I'll get started on refactoring `Printer` a little further so `Printer::apply_edit()`, `Printer::register_capability()`, and `Printer::unregister_capability()` are all changed to be `async fn` and start work on the client request routing.

**Comment by icsaszar** [comment]
@ebkalderon I made the requested changes.

> I think it should be safe to remove the make_request_old() and make_notification_old() functions completely.

I've also removed the tests that used these since without the reference values, there isn't much we can test.

**Comment by ebkalderon** [comment]
Looks terrific, thank you!

**Pull request merged**